File: app/src/main/java/de/christinecoenen/code/zapp/utils/system/NetworkConnectionHelper.java
package de.christinecoenen.code.zapp.utils.system;


import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;

import androidx.core.net.ConnectivityManagerCompat;

import java.lang.ref.WeakReference;


public class NetworkConnectionHelper {

	private final NetworkReceiver networkReceiver = new NetworkReceiver();
	private final WeakReference<Context> contextReference;
	private Listener listener;

	public NetworkConnectionHelper(Context context) {
		contextReference = new WeakReference<>(context);
	}

	public void startListenForNetworkChanges(Listener listener) {
		Context context = contextReference.get();
		if (context == null) {
			return;
		}

		this.listener = listener;
		IntentFilter filter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);
		context.registerReceiver(networkReceiver, filter);
	}

	public void endListenForNetworkChanges() {
		Context context = contextReference.get();
		if (context == null) {
			return;
		}

		this.listener = null;
		context.unregisterReceiver(networkReceiver);
	}

	public boolean isConnectedToUnmeteredNetwork() {
		Context context = contextReference.get();
		if (context == null) {
			return false;
		}

		ConnectivityManager connectionManager = (ConnectivityManager)
			context.getSystemService(Context.CONNECTIVITY_SERVICE);


		return !ConnectivityManagerCompat.isActiveNetworkMetered(connectionManager);
	}


	private class NetworkReceiver extends BroadcastReceiver {

		@Override
		public void onReceive(Context context, Intent intent) {
			if (listener != null) {
				listener.onNetworkChanged();
			}
		}
	}

	public interface Listener {
		void onNetworkChanged();
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/utils/system/NetworkConnectionHelper.kt
package de.christinecoenen.code.zapp.utils.system

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Handler
import android.os.Looper
import androidx.core.net.ConnectivityManagerCompat

/**
 * Detects changes in network status (metered or unmetered).
 * Call startListenForNetworkChanges to get callbacks whenever isConnectedToUnmeteredNetwork
 * changes.
 * Listeners will be called on main thread.
 */
class NetworkConnectionHelper(context: Context) {

	var isConnectedToUnmeteredNetwork: Boolean = true
		private set

	private var wasConnectedToUnmeteredNetwork = false

	private val mainThreadHandler = Handler(Looper.getMainLooper())
	private val connectivityManager =
		context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
	private var networkChangedCallback: (() -> Unit)? = null

	private val networkCallback = object : ConnectivityManager.NetworkCallback() {
		override fun onAvailable(network: Network) {
			onNetworkChanged()
		}

		override fun onLost(network: Network) {
			onNetworkChanged()
		}

		override fun onUnavailable() {
			onNetworkChanged()
		}

		override fun onCapabilitiesChanged(
			network: Network,
			networkCapabilities: NetworkCapabilities
		) {
			onNetworkChanged()
		}
	}

	fun startListenForNetworkChanges(networkChangedCallback: () -> Unit) {
		this.networkChangedCallback = networkChangedCallback

		connectivityManager.registerNetworkCallback(
			NetworkRequest.Builder().build(),
			networkCallback
		)
	}

	fun endListenForNetworkChanges() {
		networkChangedCallback = null
		connectivityManager.unregisterNetworkCallback(networkCallback)
	}

	private fun onNetworkChanged() {
		this.isConnectedToUnmeteredNetwork =
			!ConnectivityManagerCompat.isActiveNetworkMetered(connectivityManager)

		if (isConnectedToUnmeteredNetwork != wasConnectedToUnmeteredNetwork) {
			wasConnectedToUnmeteredNetwork = this.isConnectedToUnmeteredNetwork

			mainThreadHandler.post {
				networkChangedCallback?.invoke()
			}
		}
	}
}
